PowerPointLabs/PowerPointLabs/ActionFramework/Common/Extension/PowerpointExtension.cs

```diff
@@ -0,0 +1,192 @@
+            {
+                shape.Cut();
+                return shapes.PasteSpecial(PpPasteDataType.ppPastePNG)[1];
+            });
+        }
+
+        public static Shape SafeCut(this Shapes shapes, ShapeRange selection)
+        {
+            return PPLClipboard.Instance.LockAndRelease(() =>
+            {
+                selection.Cut();
+                return shapes.PasteSpecial(PpPasteDataType.ppPastePNG)[1];
+            });
+        }
+
+        public static Shape SafeCopyPNG(this Shapes shapes, Shape shape)
+        {
+            return PPLClipboard.Instance.LockAndRelease(() =>
+            {
+                shape.Copy();
+                return shapes.PasteSpecial(PpPasteDataType.ppPastePNG)[1];
+            });
+        }
+
+        public static void CopyPropertiesAndFieldsFrom<T>(this T target, T source)
+        {
+            target.CopyPropertiesFrom(source);
+            target.CopyFieldsFrom(source);
+        }
+
+        // copies properties recursively. Does not detect loops
+        public static void CopyPropertiesFrom<T>(this T target, T source)
+        {
+            typeof(T).CopyProperties(target, source);
+        }
+
+        public static void CopyFieldsFrom<T>(this T target, T source)
+        {
+            typeof(T).CopyFields(target, source);
+        }
+
+        public static bool IsEmptyPlaceholder(this Shape shape)
+        {
+            return shape.Type == Microsoft.Office.Core.MsoShapeType.msoPlaceholder &&
+                shape.TextFrame.TextRange.Text.Length == 0;
+        }
+
+        private static void CopyPropertiesAndFields(this Type t, object target, object source)
+        {
+            t.CopyProperties(target, source);
+            t.CopyFields(target, source);
+        }
+
+        private static void CopyProperties(this Type type, object target, object source)
+        {
+            foreach (PropertyInfo i in type.GetProperties())
+            {
+                Type t = i.PropertyType;
+                if (!i.CanRead)
+                {
+                    continue; // can't copy at all
+                }
+                else if (!i.CanWrite)
+                {
+                    // recursive copy
+                    TryAndCatch(() => t.CopyPropertiesAndFields(i.GetValue(target), i.GetValue(source)));
+                    continue;
+                }
+                // direct copy
+                TryAndCatch(() => i.SetValue(target, i.GetValue(source)));
+            }
+        }
+
+        private static void CopyFields(this Type t, object target, object source)
+        {
+            foreach (FieldInfo i in t.GetFields())
+            {
+                TryAndCatch(() => i.SetValue(target, i.GetValue(source)));
+            }
+        }
+
+        private static void TryAndCatch(Action a)
+        {
+            try
+            {
+                a();
+            }
+            catch
+            {
+                // do nothing
```

Comment:
Any reason for catching without any logging?

Reply:
When doing reflection, exceptions can occur quite often. I will add the logging